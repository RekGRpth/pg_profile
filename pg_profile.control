# Profiler extension for PostgreSQL
comment = 'PostgreSQL load profile repository and report builder'
default_version = '0.0.7'
relocatable = false
requires = 'dblink,plpgsql'
