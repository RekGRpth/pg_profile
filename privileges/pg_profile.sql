--<extension_start>
GRANT USAGE ON SCHEMA @extschema@ TO public;
--<extension_end>
GRANT SELECT ON sample_stat_cluster TO public;
GRANT SELECT ON sample_stat_slru TO public;
GRANT SELECT ON sample_stat_wal TO public;
GRANT SELECT ON sample_stat_io TO public;
GRANT SELECT ON sample_stat_archiver TO public;
GRANT SELECT ON indexes_list TO public;
GRANT SELECT ON sample_stat_indexes TO public;
GRANT SELECT ON sample_stat_indexes_total TO public;
GRANT SELECT ON tablespaces_list TO public;
GRANT SELECT ON sample_stat_tablespaces TO public;
GRANT SELECT ON tables_list TO public;
GRANT SELECT ON sample_stat_tables TO public;
GRANT SELECT ON sample_stat_tables_total TO public;
GRANT SELECT ON sample_settings TO public;
GRANT SELECT ON funcs_list TO public;
GRANT SELECT ON sample_stat_user_functions TO public;
GRANT SELECT ON sample_stat_user_func_total TO public;
GRANT SELECT ON sample_stat_database TO public;
GRANT SELECT ON sample_statements TO public;
GRANT SELECT ON sample_statements_total TO public;
GRANT SELECT ON sample_kcache TO public;
GRANT SELECT ON sample_kcache_total TO public;
GRANT SELECT ON roles_list TO public;
GRANT SELECT ON wait_sampling_total TO public;
GRANT SELECT (server_id, server_name, server_description, server_created, db_exclude,
  enabled, max_sample_age, last_sample_id, size_smp_wnd_start, size_smp_wnd_dur, size_smp_interval)
  ON servers TO public;
GRANT SELECT ON samples TO public;
GRANT SELECT ON baselines TO public;
GRANT SELECT ON bl_samples TO public;
GRANT SELECT ON report_static TO public;
GRANT SELECT ON report TO public;
GRANT SELECT ON report_struct TO public;
GRANT SELECT ON extension_versions TO public;
GRANT SELECT ON table_storage_parameters TO public;
GRANT SELECT ON index_storage_parameters TO public;
GRANT SELECT ON v_sample_stat_indexes TO public;
GRANT SELECT ON v_sample_stat_tablespaces TO public;
GRANT SELECT ON v_sample_timings TO public;
GRANT SELECT ON v_sample_stat_tables TO public;
GRANT SELECT ON v_sample_settings TO public;
GRANT SELECT ON v_sample_stat_user_functions TO public;
GRANT SELECT ON v_extension_versions TO public;
GRANT SELECT ON v_table_storage_parameters TO public;
GRANT SELECT ON v_index_storage_parameters TO public;

-- pg_read_all_stats can see the query texts
GRANT SELECT ON stmt_list TO pg_read_all_stats;
