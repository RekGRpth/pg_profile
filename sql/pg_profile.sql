CREATE SCHEMA IF NOT EXISTS profile;
CREATE SCHEMA IF NOT EXISTS dblink;
CREATE SCHEMA IF NOT EXISTS statements;
CREATE EXTENSION dblink SCHEMA dblink;
CREATE EXTENSION pg_stat_statements SCHEMA statements;
CREATE EXTENSION pg_profile SCHEMA profile;

/* == Testing server management functions == */
SELECT profile.create_server('srvtest','dbname=postgres host=localhost port=5432');
SELECT * FROM servers;
SELECT profile.rename_server('srvtest','srvtestrenamed');
SELECT profile.set_server_connstr('srvtestrenamed','dbname=postgres host=localhost port=5433');
SELECT profile.set_server_db_exclude('srvtestrenamed',ARRAY['db1','db2','db3']);
SELECT profile.set_server_max_sample_age('srvtestrenamed',3);
SELECT * FROM servers;
SELECT profile.disable_server('srvtestrenamed');
SELECT * FROM servers;
SELECT profile.enable_server('srvtestrenamed');
SELECT * FROM servers;
SELECT * FROM profile.drop_server('srvtestrenamed');
/* Testing sample creation */
SELECT * FROM profile.take_sample();
SELECT * FROM profile.take_sample();
/* Testing report */
SELECT count(1) FROM profile.get_report(1,2);
/* Testing diffreport */
SELECT * FROM profile.take_sample();
SELECT count(1) FROM profile.get_diffreport(1,2,2,3);
SELECT * FROM profile.take_sample();
SELECT * FROM profile.take_sample();
UPDATE profile.samples SET sample_time = now() - '4 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 1;
UPDATE profile.samples SET sample_time = now() - '3 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 2;
UPDATE profile.samples SET sample_time = now() - '2 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 3;
UPDATE profile.samples SET sample_time = now() - '1 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 4;
UPDATE profile.samples SET sample_time = now() - '23 hours'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 5;
SELECT * FROM profile.take_sample();
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
SELECT * FROM profile.set_server_max_sample_age('local',1);
/* Testing baseline creation */
SELECT * FROM profile.create_baseline('testline1',2,4);
SELECT * FROM profile.create_baseline('testline2',2,4);
SELECT count(*) FROM baselines;
SELECT * FROM profile.keep_baseline('testline2',-1);
/* Testing baseline show */
SELECT count(*) FROM profile.show_baselines();
/* Testing baseline deletion */
SELECT * FROM profile.take_sample();
SELECT count(*) FROM baselines;
/* Testing samples retention override with baseline */
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
SELECT * FROM profile.drop_baseline('testline1');
/* Testing samples deletion after baseline removed */
SELECT * FROM profile.take_sample();
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
/* Testing drop server with data */
SELECT * FROM profile.drop_server('local');
DROP EXTENSION pg_profile;
DROP EXTENSION IF EXISTS pg_stat_statements;
DROP EXTENSION IF EXISTS dblink;
DROP SCHEMA profile;
DROP SCHEMA dblink;
DROP SCHEMA statements;
