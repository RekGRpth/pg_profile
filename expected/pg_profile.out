CREATE SCHEMA IF NOT EXISTS profile;
CREATE SCHEMA IF NOT EXISTS dblink;
CREATE SCHEMA IF NOT EXISTS statements;
CREATE EXTENSION dblink SCHEMA dblink;
CREATE EXTENSION pg_stat_statements SCHEMA statements;
CREATE EXTENSION pg_profile SCHEMA profile;
/* == Testing server management functions == */
SELECT profile.create_server('srvtest','dbname=postgres host=localhost port=5432');
 create_server 
---------------
             2
(1 row)

SELECT * FROM servers;
 server_id | server_name | db_exclude | enabled |                 connstr                  | max_sample_age | last_sample_id 
-----------+-------------+------------+---------+------------------------------------------+----------------+----------------
         1 | local       |            | t       | dbname=contrib_regression port=5432      |                |              0
         2 | srvtest     |            | t       | dbname=postgres host=localhost port=5432 |                |              0
(2 rows)

SELECT profile.rename_server('srvtest','srvtestrenamed');
 rename_server 
---------------
             1
(1 row)

SELECT profile.set_server_connstr('srvtestrenamed','dbname=postgres host=localhost port=5433');
 set_server_connstr 
--------------------
                  1
(1 row)

SELECT profile.set_server_db_exclude('srvtestrenamed',ARRAY['db1','db2','db3']);
 set_server_db_exclude 
-----------------------
                     1
(1 row)

SELECT profile.set_server_max_sample_age('srvtestrenamed',3);
 set_server_max_sample_age 
---------------------------
                         1
(1 row)

SELECT * FROM servers;
 server_id |  server_name   |  db_exclude   | enabled |                 connstr                  | max_sample_age | last_sample_id 
-----------+----------------+---------------+---------+------------------------------------------+----------------+----------------
         1 | local          |               | t       | dbname=contrib_regression port=5432      |                |              0
         2 | srvtestrenamed | {db1,db2,db3} | t       | dbname=postgres host=localhost port=5433 |              3 |              0
(2 rows)

SELECT profile.disable_server('srvtestrenamed');
 disable_server 
----------------
              1
(1 row)

SELECT * FROM servers;
 server_id |  server_name   |  db_exclude   | enabled |                 connstr                  | max_sample_age | last_sample_id 
-----------+----------------+---------------+---------+------------------------------------------+----------------+----------------
         1 | local          |               | t       | dbname=contrib_regression port=5432      |                |              0
         2 | srvtestrenamed | {db1,db2,db3} | f       | dbname=postgres host=localhost port=5433 |              3 |              0
(2 rows)

SELECT profile.enable_server('srvtestrenamed');
 enable_server 
---------------
             1
(1 row)

SELECT * FROM servers;
 server_id |  server_name   |  db_exclude   | enabled |                 connstr                  | max_sample_age | last_sample_id 
-----------+----------------+---------------+---------+------------------------------------------+----------------+----------------
         1 | local          |               | t       | dbname=contrib_regression port=5432      |                |              0
         2 | srvtestrenamed | {db1,db2,db3} | t       | dbname=postgres host=localhost port=5433 |              3 |              0
(2 rows)

SELECT * FROM profile.drop_server('srvtestrenamed');
 drop_server 
-------------
           1
(1 row)

/* Testing sample creation */
SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

/* Testing report */
SELECT count(1) FROM profile.get_report(1,2);
 count 
-------
     1
(1 row)

/* Testing diffreport */
SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT count(1) FROM profile.get_diffreport(1,2,2,3);
NOTICE:  relation "queries_list" already exists, skipping
 count 
-------
     1
(1 row)

SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

UPDATE profile.samples SET sample_time = now() - '4 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 1;
UPDATE profile.samples SET sample_time = now() - '3 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 2;
UPDATE profile.samples SET sample_time = now() - '2 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 3;
UPDATE profile.samples SET sample_time = now() - '1 days'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 4;
UPDATE profile.samples SET sample_time = now() - '23 hours'::interval - '10 minutes'::interval WHERE server_id = 1 AND sample_id = 5;
SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
 count 
-------
     4
(1 row)

SELECT * FROM profile.set_server_max_sample_age('local',1);
 set_server_max_sample_age 
---------------------------
                         1
(1 row)

/* Testing baseline creation */
SELECT * FROM profile.create_baseline('testline1',2,4);
 create_baseline 
-----------------
               1
(1 row)

SELECT * FROM profile.create_baseline('testline2',2,4);
 create_baseline 
-----------------
               2
(1 row)

SELECT count(*) FROM baselines;
 count 
-------
     2
(1 row)

SELECT * FROM profile.keep_baseline('testline2',-1);
 keep_baseline 
---------------
             1
(1 row)

/* Testing baseline show */
SELECT count(*) FROM profile.show_baselines();
 count 
-------
     2
(1 row)

/* Testing baseline deletion */
SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT count(*) FROM baselines;
 count 
-------
     1
(1 row)

/* Testing samples retention override with baseline */
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
 count 
-------
     3
(1 row)

SELECT * FROM profile.drop_baseline('testline1');
 drop_baseline 
---------------
             1
(1 row)

/* Testing samples deletion after baseline removed */
SELECT * FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
 count 
-------
     0
(1 row)

/* Testing drop server with data */
SELECT * FROM profile.drop_server('local');
 drop_server 
-------------
           1
(1 row)

DROP EXTENSION pg_profile;
DROP EXTENSION IF EXISTS pg_stat_statements;
DROP EXTENSION IF EXISTS dblink;
DROP SCHEMA profile;
DROP SCHEMA dblink;
DROP SCHEMA statements;
